packages/cli/src/ui/hooks/useHistoryManager.ts

```diff
@@ -80,26 +100,38 @@ export function useHistory(): UseHistoryManagerReturn {
       id: number,
       updates: Partial<Omit<HistoryItem, 'id'>> | HistoryItemUpdater,
     ) => {
+      let updatedItem: HistoryItem | undefined;
       setHistory((prevHistory) =>
         prevHistory.map((item) => {
           if (item.id === id) {
             // Apply updates based on whether it's an object or a function
             const newUpdates =
               typeof updates === 'function' ? updates(item) : updates;
-            return { ...item, ...newUpdates } as HistoryItem;
+            updatedItem = { ...item, ...newUpdates } as HistoryItem;
+            return updatedItem;
           }
           return item;
         }),
       );
+      if (updatedItem) {
+        onItemUpdated?.(updatedItem);
+      }
     },
-    [],
+    [onItemUpdated],
   );
 
   // Clears the entire history state and resets the ID counter.
   const clearItems = useCallback(() => {
-    setHistory([]);
+    let hadItems = false;
+    setHistory((prevHistory) => {
+      hadItems = prevHistory.length > 0;
+      return [];
+    });
     messageIdCounterRef.current = 0;
-  }, []);
+    if (hadItems) {
```

Comment:
having side effects within set callbacks insn't a great practice in react. I'd suggest we simplify and just always log that the history was cleared even if it was empty and same for adding items. Don't think we will care about the difference in accuracy and would make the code have simpler timing with lower risks of possible gotchas.